plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryQueryPageSource.java

```diff
@@ -152,7 +161,21 @@ public long getMemoryUsage()
     public Page getNextPage()
     {
         verify(pageBuilder.isEmpty());
-        for (FieldValueList record : getFutureValue(tableResult).iterateAll()) {
+        TableResult tableResult = getFutureValue(tableResultFuture);
+        if (tableResult.hasNextPage()) {
+            tableResultFuture = CompletableFuture.supplyAsync(() -> {
+                long start = System.nanoTime();
+                TableResult result = tableResult.getNextPage();
+                readTimeNanos.addAndGet(System.nanoTime() - start);
+                return result;
+            }, executor);
```

Comment:
So it's ok if `getNextPage()` is blocking, but we should avoid blocking only in `isFinished()`? I got confused by https://github.com/trinodb/trino/pull/23741#discussion_r1800510989 where you wrote:
> So we should use futures for next calls rather than for the creation of the iterator and isFinished should be changed to responses.isReady() && !responses.hasNext()

I guess we could use the `com.google.api.gax.rpc.ServerStream#isReceiveReady` method, but as:
```
    @Override
    public boolean isFinished()
    {
        return !responses.isReady() || !responses.hasNext();
    }

```

Reply:
I made it consistent with changes in the Arrow and Avro page sources, where we now prefetch the next response.

Reply:
Yes, I think we should keep the triggering of actual work lazy. It's okay for getNextPage to return a null until it's ready to provide next page and meanwhile isBlocked can tell the engine when the connector will be ready with next page.

Reply:
Doesn't returning `null` signal there's no more pages? The docs are not clear on this.

Reply:
no, `isFinished` is the source of truth for whether there are more pages, engine isn't supposed to infer anything from a `null` page

Reply:
I remove the usage of futures except for scheduling it once in the constructor.